torch/cuda/__init__.py

```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+    cur_stream : Optional['torch.classes.cuda.Stream']  # type: ignore
+    cur_stream_device: Optional[_device]
+
+    def __init__(self, stream: Optional['torch.classes.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
+        # Initialize the below streams to default stream on the current device
+        self.device_index = _get_device_index(device=None, optional=True)
+        self.src_prev_stream = torch.cuda.default_stream(self.device_index)
+        self.dst_prev_stream = torch.cuda.default_stream(self.device_index)
 
+    def __enter__(self):
+        self.idx = _get_device_index(device=None, optional=True)
+        # If there is no CUDA device available, return
+        if self.idx == -1:
+            return
+
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # Return if stream is None
+        if cur_stream is None:
+            return
+        self.src_prev_stream = torch.cuda.current_stream(self.idx)
+
+        # Eager mode and script mode APIs differ a bit. Hence, we need to check
+        # if we are in scripting mode or eager mode. Depending on the mode
+        # we can call the corresponding API's to get the device index.
+        if torch.jit.is_scripting():
+            self.prev_stream_device_index = self.src_prev_stream.device_index()  # type: ignore
+            self.cur_stream_device_index = cur_stream.device_index()
+        else:
+            self.prev_stream_device_index = _get_device_index(self.src_prev_stream.device, optional=True)
+            self.cur_stream_device_index = _get_device_index(cur_stream.device, optional=True)
+
+        # If the stream is not on the current device, then change the device
+        # and set the current stream on the device
+        if self.prev_stream_device_index != self.cur_stream_device_index:
+            # Check if we are in eager mode or script mode and then set the
+            # current stream device by calling the corresponding API
+            if torch.jit.is_scripting():
+                cur_stream_device = cur_stream.device()
+            else:
+                cur_stream_device = cur_stream.device
+            with device(cur_stream_device):
+                self.dst_prev_stream = torch.cuda.current_stream(self.cur_stream_device_index)
+
+            # The JIT and eager API to set the device differ. Hence, call the
+            # correct API by checking the current mode (eager/JIT).
+            if torch.jit.is_scripting():
+                torch.cuda._setDevice(self.cur_stream_device_index)
+            else:
+                torch._C._cuda_setDevice(self.cur_stream_device_index)
```

Comment:
In eager mode, this code is included inside `_cuda_setStream`. Can we do the same in JIT? Can we extend the implementation of the custom JIT operator to switch devices if needed?

Reply:
```
PyObject * THCPModule_setStream_wrap(PyObject *self, PyObject *obj)
{
  HANDLE_TH_ERRORS
  THPUtils_assert(PyLong_Check(obj), "invalid stream");
  uint64_t bits = PyLong_AsUnsignedLongLong(obj);
  if (bits == static_cast<uint64_t>(-1) && PyErr_Occurred()) {
    throw python_error();
  }
  auto stream = at::cuda::CUDAStream::unpack(bits);
  auto device = static_cast<int>(c10::cuda::current_device());
  if (device != stream.device_index()) {
    THCPModule_setDevice(stream.device_index());
  }
  at::cuda::setCurrentCUDAStream(stream);
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

in `torch/csrc/cuda/Module.cpp` for reference.

Reply:
OK agreed, fixed this